daemon/control_ng.c

```diff
@@ -62,6 +63,104 @@ const char *ng_command_strings_short[NGC_COUNT] = {
 	"Pub", "SubReq", "SubAns", "Unsub",
 };
 
+typedef struct ng_ctx {
+	str callid;
+	str command;
+	str cookie;
+	bool should_trace;
+	const endpoint_t *sin_ep;
+	const endpoint_t *local_ep;
+} ng_ctx;
+
+#define CH(func, ...) do { \
+	if (rtpe_config.homer_ng) \
+		func( __VA_ARGS__); \
+} while (0)
+
+static GString *create_homer_msg(str *cookie, str *data) {
+	GString *msg = g_string_sized_new(cookie->len + 1 + data->len);
+	g_string_append_printf(msg, "%.*s %.*s", STR_FMT(cookie), STR_FMT(data));
+	return msg;
+}
+
+static bool should_trace_msg(str *pcmd) {
+	switch (__csh_lookup(pcmd)) {
+		case CSH_LOOKUP("ping"):
+			return false;
+		default:
+			return true;
+	}
+}
+
+static void homer_fill_values(ng_ctx *hctx, str *callid, str *cmd) {
+	if (hctx) {
+		hctx->command = *cmd;
+		hctx->callid = *callid;
+	}
+}
+
+static void homer_extract_values(ng_ctx *hctx, str* data, struct obj *ref, struct ng_buffer **ngbufp) {
+	bencode_item_t *dict, *resp;
+
+	struct ng_buffer *ngbuf = *ngbufp = ng_buffer_new(ref);
+
+	resp = bencode_dictionary(&ngbuf->buffer);
+	assert(resp != NULL);
+
+	//"Invalid data (no payload)";
+	if (data->len <= 0)
+		goto end_function;
+
+	if (data->s[0] == 'd') {
+		dict = bencode_decode_expect_str(&ngbuf->buffer, data, BENCODE_DICTIONARY);
+		//"Could not decode bencode dictionary";
+		if (!dict)
+			goto end_function;
+	}
+	else if (data->s[0] == '{') {
+		JsonParser *json = json_parser_new();
+		bencode_buffer_destroy_add(&ngbuf->buffer, g_object_unref, json);
+		//"Failed to parse JSON document";
+		if (!json_parser_load_from_data(json, data->s, data->len, NULL))
+			goto end_function;
+		dict = bencode_convert_json(&ngbuf->buffer, json);
+		//"Could not decode bencode dictionary";
+		if (!dict || dict->type != BENCODE_DICTIONARY)
+			goto end_function;
+	}
+	else {
+		//"Invalid NG data format";
```

Comment:
Any way to avoid duplicating this code? Is this even needed? AFAICT it's only used for sending cached replies. Cached replies are retransmits by definition. Are these of interest to the Homer backend?

Reply:
Hi, I think that in debugging scenarios one would be interested to see that retransmissions are sent. To avoid duplication of code maybe I can also cache some values just as the reply itself is cached.